src/kakarot/state.cairo

```diff
@@ -0,0 +1,518 @@
+        }
+    }
+
+    // @notice Update a storage key with the given value
+    // @param self The pointer to the State.
+    // @param address The pointer to the Account address
+    // @param key The pointer to the Uint256 storage key
+    // @param value The pointer to the Uint256 value
+    func write_storage{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr}(
+        self: model.State*, address: model.Address*, key: Uint256*, value: Uint256*
+    ) -> model.State* {
+        alloc_locals;
+        let (self, account) = get_account(self, address);
+        local storage: DictAccess* = account.storage;
+        let (storage_key) = hash_felts{hash_ptr=pedersen_ptr}(cast(key, felt*), 2);
+        dict_write{dict_ptr=storage}(key=storage_key, new_value=cast(value, felt));
+        tempvar account = new model.Account(
+            account.code_len, account.code, account.storage_start, storage, account.nonce
+        );
+        let self = set_account(self, address, account);
+        return self;
+    }
+
+    // @notice Add an event to the Event* array
+    // @param self The pointer to the State
+    // @param event The pointer to the Event
+    // @return The updated State
+    func add_event(self: model.State*, event: model.Event*) -> model.State* {
+        assert [self.events + self.events_len] = [event];
+        return new model.State(
+            accounts_start=self.accounts_start,
+            accounts=self.accounts,
+            events_len=self.events_len + 1,
+            events=self.events,
+            balances_start=self.balances_start,
+            balances=self.balances,
+            transfers_len=self.transfers_len,
+            transfers=self.transfers,
+        );
+    }
+
+    // @notice Add a transfer to the Transfer* array
+    // @param self The pointer to the State
+    // @param event The pointer to the Transfer
+    // @return The updated State
+    func add_transfer{range_check_ptr}(
+        self: model.State*, transfer: model.Transfer*
+    ) -> model.State* {
+        alloc_locals;
+        // See https://docs.cairo-lang.org/0.12.0/how_cairo_works/functions.html#retrieving-registers
+        let fp_and_pc = get_fp_and_pc();
+        local __fp__: felt* = fp_and_pc.fp_val;
+
+        let balances = self.balances;
+
+        let (local pointer) = dict_read{dict_ptr=balances}(transfer.sender.evm);
+        tempvar sender_balance_prev_ptr = cast(pointer, Uint256*);
+        tempvar sender_balance_prev = Uint256(
+            sender_balance_prev_ptr.low, sender_balance_prev_ptr.high
+        );
+
+        let (local pointer) = dict_read{dict_ptr=balances}(transfer.recipient.evm);
+        tempvar recipient_balance_prev_ptr = cast(pointer, Uint256*);
+        tempvar recipient_balance_prev = Uint256(
+            recipient_balance_prev_ptr.low, recipient_balance_prev_ptr.high
+        );
+
+        let (local sender_balance_new) = uint256_sub(sender_balance_prev, transfer.amount);
```

Comment:
won't this wrap around if sender_balance_prev < transfer.amount?

Reply:
good catch

Reply:
actually probably at the ExecutionContext level and not at the State, I'll figure out

Reply:
done